service/interpreter/timers/greedyTimerProcessor.go

```diff
@@ -0,0 +1,264 @@
+				fireAt := next.FiringUnixTimestampSeconds
+				duration := time.Duration(fireAt-now) * time.Second
+				t.provider.NewTimer(ctx, duration)
+				createdTimers = append(createdTimers, fireAt)
+			}
+		}
+	})
+}
+
+func (t *GreedyTimerProcessor) Dump() []service.StaleSkipTimerSignal {
+	return t.staleSkipTimerSignals
+}
+
+func (t *GreedyTimerProcessor) GetCurrentTimerInfos() map[string][]*service.TimerInfo {
+	return t.stateExecutionCurrentTimerInfos
+}
+
+// SkipTimer will attempt to skip a timer, return false if no valid timer found
+func (t *GreedyTimerProcessor) SkipTimer(stateExeId, timerId string, timerIdx int) bool {
+	timer, valid := service.ValidateTimerSkipRequest(t.stateExecutionCurrentTimerInfos, stateExeId, timerId, timerIdx)
+	if !valid {
+		// since we have checked it before sending signals, this should only happen in some vary rare cases for racing condition
+		t.logger.Warn("cannot process timer skip request, maybe state is already closed...putting into a stale skip timer queue", stateExeId, timerId, timerIdx)
+
+		t.staleSkipTimerSignals = append(t.staleSkipTimerSignals, service.StaleSkipTimerSignal{
+			StateExecutionId:  stateExeId,
+			TimerCommandId:    timerId,
+			TimerCommandIndex: timerIdx,
+		})
+		return false
+	}
+	timer.Status = service.TimerSkipped
+	return true
+}
+
+func (t *GreedyTimerProcessor) RetryStaleSkipTimer() bool {
+	for i, staleSkip := range t.staleSkipTimerSignals {
+		found := t.SkipTimer(staleSkip.StateExecutionId, staleSkip.TimerCommandId, staleSkip.TimerCommandIndex)
+		if found {
+			newList := removeElement(t.staleSkipTimerSignals, i)
+			t.staleSkipTimerSignals = newList
+			return true
+		}
+	}
+	return false
+}
+
+// WaitForTimerFiredOrSkipped waits for timer completed(fired or skipped),
+// return true when the timer is fired or skipped
+// return false if the waitingCommands is canceled by cancelWaiting bool pointer(when the trigger type is completed, or continueAsNew)
+func (t *GreedyTimerProcessor) WaitForTimerFiredOrSkipped(
+	ctx interfaces.UnifiedContext, stateExeId string, timerIdx int, cancelWaiting *bool,
+) service.InternalTimerStatus {
+	timerInfos := t.stateExecutionCurrentTimerInfos[stateExeId]
+	if len(timerInfos) == 0 {
+		if *cancelWaiting {
+			// The waiting thread is later than the timer execState thread
+			// The execState thread got completed early and call RemovePendingTimersOfState to remove the timerInfos
+			// returning pending here
+			return service.TimerPending
+		} else {
+			panic("bug: this shouldn't happen")
+		}
+	}
+	timer := timerInfos[timerIdx]
+	if timer.Status == service.TimerFired || timer.Status == service.TimerSkipped {
+		return timer.Status
+	}
+	skippedByStaleSkip := t.RetryStaleSkipTimer()
+	if skippedByStaleSkip {
+		t.logger.Warn("timer skipped by stale skip signal", stateExeId, timerIdx)
+		timer.Status = service.TimerSkipped
+		return service.TimerSkipped
+	}
+
+	_ = t.provider.Await(ctx, func() bool {
```

Comment:
Nit: add a comment here that the timers in the TimerScheduler will trigger a workflow task when the timer fire, hence it will drive the workflow task to evalaute `timer.FiringUnixTimestampSeconds <= t.provider.Now(ctx).Unix()`. 

That was a tricky part too in our Wayfinder code :P